vita3k/util/src/cpu_detect.cpp

```diff
@@ -0,0 +1,234 @@
+// Code adapted from https://github.com/chronoxor/CppCommon
+/*!
+    \file cpu.cpp
+    \brief CPU management implementation
+    \author Ivan Shynkarenka
+    \date 27.07.2016
+    \copyright MIT License
+*/
+
+#include "util/cppcommon.h"
+#include "util/cppcommon_resource.h"
+
+#if defined(__APPLE__)
+#include <sys/sysctl.h>
+#elif defined(unix) || defined(__unix) || defined(__unix__)
+#include <fstream>
+#include <regex>
+#include <unistd.h>
+#elif defined(_WIN32) || defined(_WIN64)
+#include <windows.h>
+#endif
+
+namespace CppCommon {
+
+//! @cond INTERNALS
+namespace Internals {
+
+#if defined(_WIN32) || defined(_WIN64)
+
+// Helper function to count set bits in the processor mask
+DWORD CountSetBits(ULONG_PTR pBitMask) {
+    DWORD dwLeftShift = sizeof(ULONG_PTR) * 8 - 1;
+    DWORD dwBitSetCount = 0;
+    ULONG_PTR pBitTest = (ULONG_PTR)1 << dwLeftShift;
+
+    for (DWORD i = 0; i <= dwLeftShift; ++i) {
+        dwBitSetCount += ((pBitMask & pBitTest) ? 1 : 0);
+        pBitTest /= 2;
+    }
+
+    return dwBitSetCount;
+}
+
+#endif
+
+} // namespace Internals
+//! @endcond
+
+std::string CPU::Architecture() {
+#if defined(__APPLE__)
+    char result[1024];
+    size_t size = sizeof(result);
+    if (sysctlbyname("machdep.cpu.brand_string", result, &size, nullptr, 0) == 0)
+        return result;
+
+    return "<unknown>";
+#elif defined(unix) || defined(__unix) || defined(__unix__)
+    static std::regex pattern("model name(.*): (.*)");
+
+    std::string line;
+    std::ifstream stream("/proc/cpuinfo");
+    while (getline(stream, line)) {
+        std::smatch matches;
+        if (std::regex_match(line, matches, pattern))
+            return matches[2];
+    }
+
+    return "<unknown>";
+#elif defined(_WIN32) || defined(_WIN64)
+    HKEY hKeyProcessor;
+    LONG lError = RegOpenKeyExA(HKEY_LOCAL_MACHINE, "HARDWARE\\DESCRIPTION\\System\\CentralProcessor\\0", 0, KEY_READ, &hKeyProcessor);
+    if (lError != ERROR_SUCCESS)
+        return "<unknown>";
+
+    // Smart resource cleaner pattern
+    auto key = resource(hKeyProcessor, [](HKEY hKey) { RegCloseKey(hKey); });
+
+    CHAR pBuffer[_MAX_PATH] = { 0 };
+    DWORD dwBufferSize = sizeof(pBuffer);
+    lError = RegQueryValueExA(key.get(), "ProcessorNameString", nullptr, nullptr, (LPBYTE)pBuffer, &dwBufferSize);
+    if (lError != ERROR_SUCCESS)
+        return "<unknown>";
+
+    // Remove trailing 0x20 space from pBuffer
+    while (!std::isgraph(pBuffer[dwBufferSize - 1]) && dwBufferSize > 0) {
+        pBuffer[--dwBufferSize] = '\0';
+    }
+
+    return std::string(pBuffer);
+#else
+#error Unsupported platform
```

Comment:
I'm not sure if android is linux, but maybe android also should be supported. At least don't crash on android.

Reply:
Android is indeed linux (and both __linux and __linux__ are defined) so it's fine